### PR TITLE
MSL: Add support for SPV_KHR_physical_storage_buffer extension.

### PR DESCRIPTION
**Pull request by billhollings**
- Determine sizing and alignments of pointers to types as distinct from the size and alignment of the types themselves.
- Declare all buffer pointers in the MSL device address space.
- Support struct pointer recursion, where structs can contain pointers to themselves or to a parent struct.
- Add SPIRType::was_forward_referenced to track if a type was forward referenced, to help emit MSL structs in the correct dependency order.
- Handle pointers to pointers that are not just arrays of arrays.
- Add unit test shaders.

### CODE REVIEW

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Missing coverage for PtrAccessChain and OpConvertUToPtr / OpBitcast to cast integers <-> pointers.

File: spirv_msl.cpp [hunk 1]
```diff
@@ -15066,6 +15099,10 @@ uint32_t CompilerMSL::get_declared_input_size_msl(const SPIRType &type, uint32_t
 // Returns the byte alignment of a type.
 uint32_t CompilerMSL::get_declared_type_alignment_msl(const SPIRType &type, bool is_packed, bool row_major) const
 {
+	// Pointers aligns on multiples of 8 bytes
+	if (type.pointer && type.storage == StorageClassPhysicalStorageBuffer)
+		return 8 * (type.vecsize == 3 ? 4 : type.vecsize);
```
[the remark this answers is not in the document]
Comment (billhollings) [reply spirv_msl.cpp:15104]: In `ensure_member_packing_rules_msl()`, very short arrays can be converted to vectors to support wide array strides, and metal does have a `ulongn` type that supports the use of this with an array of pointers. 

The use case that triggered this was a short SPIR-V array of pointers with stride 16.

File: spirv_msl.cpp [hunk 2]
```diff
@@ -13528,17 +13543,24 @@ string CompilerMSL::type_to_glsl(const SPIRType &type, uint32_t id)
 		const char *restrict_kw;
 
 		auto type_address_space = get_type_address_space(type, id);
-		auto type_decl = type_to_glsl(get<SPIRType>(type.parent_type), id);
+		const auto* p_parent_type = &get<SPIRType>(type.parent_type);
```
[the remark this answers is not in the document]
Comment (billhollings) [reply spirv_msl.cpp:13546]: Fixed.

File: spirv_msl.cpp [hunk 3]
```diff
@@ -13528,17 +13543,24 @@ string CompilerMSL::type_to_glsl(const SPIRType &type, uint32_t id)
...
 		// but we'll have to emit thread T * thread * T constant bar; for example.
 		if (type_is_pointer_to_pointer(type))
-			type_name = join(type_decl, " ", type_address_space, " ");
+			type_name = join(type_to_glsl(*p_parent_type, id), " ", type_address_space, " ");
 		else
-			type_name = join(type_address_space, " ", type_decl);
+		{
+			// Since this is not a pointer-to-pointer, ensure we've dug down to the base type.
+			// Some situations chain pointers even though they are not formally pointers-of-pointers.
+			while(type_is_pointer(*p_parent_type))
```
[the remark this answers is not in the document]
Comment (billhollings) [reply spirv_msl.cpp:13559]: Fixed.

File: spirv_parser.cpp [hunk 4]
```diff
@@ -659,6 +659,16 @@ void Parser::parse(const Instruction &instruction)
...
 		uint32_t id = ops[0];
 
+		// The type may have previously been declared as an OpTypeForwardPointer.
+		// If so, we want to retain that knowledge, in case content needs to be emitted
+		// relative to when the forward pointer was declared, not when the pointer is
+		// declared. SPIR-V allows composites to contain a type that has only been
+		// declared as a forward pointer when the composite is defined, but in many
+		// languages, it is necessary to emit the component before the composite.
+		// Retrieve this info *before* overwriting the type info for the id below.
+		auto *fwd_ptr = maybe_get<SPIRType>(id);
```
[the remark this answers is not in the document]
Comment (billhollings) [reply spirv_parser.cpp:669]: Fixed.

**Comment by billhollings** [comment]
> Does Metal even support buffer device address?

Yes, with recently released Metal3.

> Missing coverage for PtrAccessChain and OpConvertUToPtr / OpBitcast to cast integers <-> pointers.

At first look, all of these seem to be covered with code existing prior to this PR. Can you clarify where you think these will break with added support for physical storage buffer? If we don't have specific use cases or examples for that, perhaps we can leave for a future enhancement?

Other fixes applied.

**Comment by cdavis5e** [comment]
> At first look, all of these seem to be covered with code existing prior to this PR. Can you clarify where you think these will break with added support for physical storage buffer? If we don't have specific use cases or examples for that, perhaps we can leave for a future enhancement?

Actually, that code, which I added for `SPV_KHR_variable_pointers` support, is already quite broken. It's something I intend to fix over the next six months working on MoltenVK's conformance.

**Comment by HansKristian-Work** [comment]
Yes, uvec2 -> pointer cast is broken in e.g.:

```
#version 450
#extension GL_EXT_buffer_reference : require
#extension GL_EXT_buffer_reference_uvec2 : require

layout(buffer_reference, buffer_reference_align = 4) buffer SSBO
{
	vec3 a;
};

layout(push_constant) uniform UBO
{
	uvec2 b;
};

void main()
{
	SSBO(b).a = vec3(1.0, 2.0, 3.0);
}
```

which becomes

```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct SSBO;

struct UBO
{
    uint2 b;
};

struct SSBO
{
    float3 a;
};

kernel void main0(constant UBO& _10 [[buffer(0)]])
{
    (device SSBO*(_10.b))->a = float3(1.0, 2.0, 3.0);
}
```

Another consideration is that reading a float3 from BDA might have to analyze AlignedMask in OpLoad/OpStore to determine if the struct should be packed or not. Using float3 here will break, needs packed_float3.

File: spirv_msl.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (HansKristian-Work) [reply spirv_msl.cpp:15121]: Where is this triggered? This shader is incorrect:

```
#version 450
#extension GL_EXT_buffer_reference : require
#extension GL_EXT_buffer_reference_uvec2 : require
layout(local_size_x = 1) in;

layout(buffer_reference) buffer SSBO
{
	vec4 a;
};

layout(std140, set = 0, binding = 0) uniform UBO1
{
	SSBO ssbos[4]; // Expect array stride of 16.
};

void main()
{
	ssbos[gl_WorkGroupID.x].a = vec4(4.0, 5.0, 6.0, 7.0);
}
```

```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct SSBO;

struct UBO1
{
    device SSBO* ssbos[4]; // Still array stride of 8.
};

struct SSBO
{
    float4 a;
};

constant uint3 gl_WorkGroupSize [[maybe_unused]] = uint3(1u);

kernel void main0(constant UBO1& _16 [[buffer(0)]], uint3 gl_WorkGroupID [[threadgroup_position_in_grid]])
{
    _16.ssbos[gl_WorkGroupID.x]->a = float4(4.0, 5.0, 6.0, 7.0);
}
```

File: spirv_msl.cpp [hunk 5]
```diff
@@ -15007,6 +15029,17 @@ uint32_t CompilerMSL::get_declared_struct_size_msl(const SPIRType &struct_type,
 // Returns the byte size of a struct member.
 uint32_t CompilerMSL::get_declared_type_size_msl(const SPIRType &type, bool is_packed, bool row_major) const
 {
+	// Pointers take 8 bytes each
+	if (type.pointer && type.storage == StorageClassPhysicalStorageBuffer)
+	{
+		uint32_t type_size = 8 * (type.vecsize == 3 ? 4 : type.vecsize);
+		size_t dim_cnt = type.array.size();
+		for (uint32_t dim_idx = 0; dim_idx < dim_cnt; dim_idx++)
```
[the remark this answers is not in the document]
Comment (HansKristian-Work) [reply spirv_msl.cpp:15054]: type_is_pointer() can be used to determine if the type is a true pointer and not an array. It checks parent type to make sure pointer_depth is lowered. Chasing the parent type like this is the only way to do this. I think it'd make sense to do something like (untested pseudo-code, but hopefully enough to communicate the idea):

```
if (!type.array.empty() && type.pointer)
{
uint32_t dim_idx = dim_cnt - 1;
while (!type_is_pointer(*type))
{
    // Traverse all OpTypeArray that was built on top of a pointer type.
    array_size *= get_array_size(type, dim_idx);
    dim_idx--;
    type = type->parent;
}
}
```

File: spirv_msl.cpp [hunk 6]
```diff
@@ -6792,6 +6804,25 @@ void CompilerMSL::emit_specialization_constants_and_structs()
...
 
+	// Physical storage buffer pointers can have cyclical references,
+	// so emit forward declarations of them before other structs.
+	// Ignore type_id because we want the underlying struct type from the pointer.
+	ir.for_each_typed_id<SPIRType>([&](uint32_t /* type_id */, const SPIRType &type) {
+		if (type.basetype == SPIRType::Struct &&
+			type.pointer && type.storage == StorageClassPhysicalStorageBuffer &&
+			declared_structs.count(type.self) == 0)
+		{
+			statement("struct ", to_name(type.self), ";");
```
Comment (HansKristian-Work) [inline spirv_msl.cpp:6815]: This is also probably a good time to check for required Metal version for BDA.

Comment (billhollings) [reply spirv_msl.cpp:6815]: Although CPU access to device buffer memory addresses has just been added with the Metal 3 API, and MoltenVK will include a separate test for BDA for that, within SPIRV-Cross, the basics of pointers and `ulong` values in shader buffers are already separately covered by checks in SPIRV-Cross for MSL 2 & MSL 2.3 respectively, so I don't think we need to add anything specifically for BDA as a distinct feature.

File: spirv_msl.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (billhollings) [reply spirv_msl.cpp:15054]: Fixed now with code similar to your pseudo-code.

File: spirv_msl.cpp (hunk 6 above)
Comment (HansKristian-Work) [reply spirv_msl.cpp:6815]: ok

**Comment by billhollings** [comment]
> Yes, uvec2 -> pointer cast is broken in e.g.:

> Another consideration is that reading a float3 from BDA might have to analyze AlignedMask in OpLoad/OpStore to determine if the struct should be packed or not. Using float3 here will break, needs packed_float3.

Both fixed now, and test shader added. 

MSL actually handles the assignment between `floatN` and `packed_floatN` for `OpLoad/OpStore`.

**Review by HansKristian-Work — APPROVED** [review]
LGTM. If there are further issues we can always iterate on that.

**Pull request merged**